### ISSUE TITLE
Audit forced FP16 on native-BF16 GDN composes under vLLM v0.30.0

### ISSUE THREAD

**Issue by TreyThomasCodes** [noonghunna/club-3090#1393]
## Summary

The shipped `vllm/agents-a1-dual` compose boots and serves HTTP 200 under vLLM v0.30.0, but with `--dtype float16` it generates pathological repeated punctuation (`!!!!!!!!!!!!!!!!...`) for both chat and text completion. Greedy completion also returns a vLLM 400 caused by non-JSON-compliant `NaN` logits.

Changing only `--dtype` from `float16` to `bfloat16` fixes the model. This is a silent-corruption class: health and model-list probes pass, so a readiness-only check does not catch it.

## Confirmed reproduction

Hardware: 2x RTX 3090, NVLink detected, TP=2
Image: `vllm/vllm-openai:v0.30.0`
Compose: `models/agents-a1/vllm/compose/dual/fp8-dynamic/fp8.yml`
Checkpoint: `InternScience/Agents-A1-FP8-dynamic`

Checkpoint integrity was verified against Hugging Face:

```text
model.safetensors sha256
03d3d26cc391fa006ef616e0b1ac51d5199a034f2ffd3bc9143093219fb8809d
```

FP16 behavior:

```text
POST /v1/chat/completions -> 200
content: "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
finish_reason: "length"
```

The boot log explicitly reports:

```text
Casting torch.bfloat16 to torch.float16.
Falling back to the Triton GDN decode path: the fused CUDA kernel requires a BF16 GDN model ...
```

Raw completion reproduces the repeated punctuation, independent of the chat template. `temperature=0` additionally exposes NaN logits through the JSON response error.

## Confirmed fix

One-variable A/B: change `--dtype float16` to `--dtype bfloat16`.

Results on the same container/image/checkpoint:

- `verify.sh`: 4/4 pass
- `verify-full.sh`: 10/10 pass
- Basic completion returns Paris
- Tool calling, streaming tool calls, reasoning extraction, 2K-token cascade detection, and vision 4/4 pass
- Targeted compose/catalog guards pass

## Audit candidates — not yet confirmed broken

A static audit found five other active vLLM v0.30.0 Qwen3-Next/GDN composes forcing FP16 despite native-BF16 checkpoint configs:

- `models/qwen3.6-27b/vllm/compose/single/autoround-int4/minimal.yml`
- `models/qwen3.6-27b/vllm/compose/dual/autoround-int4/fp8-mtp.yml`
- `models/qwen3.6-27b/vllm/compose/multi4/autoround-int4/mtp.yml`
- `models/qwen3.6-35b-a3b/vllm/compose/dual/autoround-int4/fp8.yml`
- `models/qwen3.6-35b-a3b/vllm/compose/single/autoround-int4/preview.yml`

These are candidates only; do not bulk-change them without one-variable live validation. For comparison, all 20 active Qwen3.8 vLLM composes use BF16, and most other active Qwen3.6 composes use BF16.

The remaining active forced-FP16 entries are Gemma composes on the separate `gemma4-unified` image and a Nemotron Mamba2/attention hybrid; they are not automatically implicated by this GDN finding.

## Suggested follow-up

1. Land the confirmed Agents-A1 one-line fix separately.
2. Run FP16/BF16 A/Bs on one representative Qwen3.6-27B and Qwen3.6-35B-A3B compose.
3. If reproduced, fix only the validated affected set.
4. Add a catalog guard against native-BF16 Qwen3-Next/GDN rows silently forcing FP16, with an explicit exception mechanism if a measured compose genuinely needs FP16.

**Comment by noonghunna** [comment]
Thanks. This is a very clean report: a one-variable A/B, a checksum-verified checkpoint, and the right boot-log lines quoted.

## What we measured

The Qwen3.6 and Agents-A1 weights aren't on the reference rig at the moment, so I tested whether **fp16 on a native-BF16 GDN model is broken on v0.30.0 in general**. The stand-in was `Qwen3.5-4B`: a Qwen3-Next GDN hybrid with native BF16 weights, TP=1, 2× 3090 PCIe. It was one fresh container per run, and only one variable changed between runs:

| Run | Chat, temp 0 | Raw completion, temp 0 | 200-token prose, temp 0.7 |
|---|---|---|---|
| v0.30.0, `--dtype float16` | `Paris` | ` Paris.\nA. True…` | coherent |
| v0.30.0, `--dtype bfloat16` | `Paris` | identical | identical |
| v0.29.0, `--dtype float16` | `Paris` | identical | identical |

All three runs are token-for-token identical at temp 0, and the prose matches too. The fp16 runs log the same two lines you quoted: `Casting torch.bfloat16 to torch.float16` and `Falling back to the Triton GDN decode path…`. **v0.29.0 logs that fallback as well**, so it isn't new in v0.30.0 and isn't the bug by itself. Casting to fp16 and running the Triton GDN decode path is healthy on v0.30.0, at least at this size and topology.

For history: Agents-A1 passed `verify-full` **at fp16** on v0.24.0 when it was onboarded (2026-07-03, PCIe TP=2).

## What that leaves, and one run that would split it

The stand-in doesn't cover TP=2, the FP8 MoE path (Marlin weight-only on sm_86), or model-specific fp16 overflow. Your NaN logits at temp 0 fit an overflow signature. It also doesn't cover something specific to **your** rig: NVLink. This compose enables vLLM's custom all-reduce **only when NVLink is detected** (`_CUSTOM_AR_ENABLED`). The reference rig is PCIe, so it has never run fp16 through custom all-reduce.

Could you run the failing arm once more, changing only that?

```bash
DISABLE_CUSTOM_ALL_REDUCE=1 bash scripts/switch.sh --force vllm/agents-a1-dual   # still --dtype float16
```

- **Clean:** the trigger is fp16 × custom all-reduce on NVLink. That isn't a GDN property at all, and it would matter for every fp16 TP≥2 compose on NVLink rigs.
- **Still `!!!!`:** it's the model or the FP8-MoE path, and your BF16 fix is the right fix.

## On your suggested follow-ups

1. **The Agents-A1 one-liner:** a PR is welcome. BF16 is the checkpoint's native dtype, and your A/B plus `verify-full` 10/10 is exactly the evidence we'd want. It's safe whichever way the run above comes out.
2. **Qwen3.6 FP16 composes:** agreed, no bulk change. The result above says the generic path is fine, so each one needs its own one-variable A/B before it's touched.
3. **A catalog guard:** let's wait for the all-reduce result. If the cause is custom all-reduce, a "no FP16 on BF16 GDN rows" guard would be aimed at the wrong variable.

**Comment by TreyThomasCodes** [comment]
Ran the requested split test. **Disabling custom all-reduce does not fix the FP16 corruption.**

Exact launch, from the unmodified `master` compose (`--dtype float16`):

```bash
DISABLE_CUSTOM_ALL_REDUCE=1 bash scripts/switch.sh --force vllm/agents-a1-dual
```

Confirmed in the container and engine logs:

```text
DISABLE_CUSTOM_ALL_REDUCE=1
custom all-reduce OFF by operator request
'dtype': 'float16'
'disable_custom_all_reduce': True
Casting torch.bfloat16 to torch.float16.
GDN decode kernel: triton
```

Results:

| Probe | Result |
|---|---|
| Chat, temp 0 | HTTP 200, `ductductduct...`, `finish_reason=length` |
| Raw completion, temp 0 | HTTP 200, `ductductduct...`, `finish_reason=length` |
| 200-token prose, temp 0.7 | HTTP 200, `!!!!!!!!!!!!!!!!...`, `finish_reason=length` |

I then restored the one-variable BF16 change and relaunched with custom all-reduce back on. The same deterministic chat probe returns:

```text
HTTP 200
finish_reason: stop
content: 'Paris'
```

So this rules out the NVLink custom-all-reduce interaction on this rig. The trigger remains specific to Agents-A1 and/or its FP8-MoE path under FP16, and the native-BF16 compose fix remains the correct scoped change.

**Comment by noonghunna** [comment]
Thanks. That's a clean split: `custom all-reduce OFF by operator request` in the log, and still garbage. So NVLink custom all-reduce is ruled out, and the trigger is Agents-A1's own fp16 path: the model's activations, and/or the FP8-MoE path (Marlin weight-only on sm_86) running in fp16. The `ductductduct` degenerate loop at temp 0, together with the NaN logits you saw earlier, fits an fp16 overflow.

On scope, this stays contained. **Agents-A1 is the only vLLM compose in the catalog that combines `--dtype float16` with FP8 weights.** The five Qwen3.6 fp16 candidates you listed are AutoRound INT4, so they don't share the path that broke. The generic "fp16 cast on a native-BF16 GDN model" path was clean in our stand-in test on both v0.29.0 and v0.30.0. So:

- **#1403 is the right fix.** It's the single-flag change, back to the checkpoint's native dtype, with your `verify-full` 10/10 behind it. We'll review it from here.
- **Qwen3.6 fp16 composes:** unchanged unless a one-variable A/B on that compose shows a problem. There's no evidence yet that they're affected.
- **Catalog guard:** not warranted on this evidence. The failure is specific to this model and quant path, not "fp16 on a BF16 GDN model", so a rule keyed on that would flag healthy composes.

One note for the record: Agents-A1 passed `verify-full` **at fp16** on v0.24.0 when it was onboarded, so something between v0.24 and v0.30 changed its fp16 behaviour. The fix doesn't depend on pinning that down, so there's no need to bisect it.
